<a href="https://colab.research.google.com/github/vnikeshana/ML_Question_06/blob/main/main_code_MLQ6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import os
import pickle
import copy
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler, LabelEncoder

import matplotlib.pyplot as plt

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:

%cd drive/MyDrive/Customs-Declaration-Datasets-en/

[Errno 2] No such file or directory: 'drive/MyDrive/Customs-Declaration-Datasets-en/'
/content


In [ ]:
class label_encoder(object):
    def fit_pd(self,df,cols=[]):
        if len(cols) == 0:
            cols = df.columns
        self.class_index = {}
        for f in cols:
            uf = df[f].unique()
            self.class_index[f] = {}
            index = 1
            for item in uf:
                self.class_index[f][item] = index
                index += 1

    def fit_transform_pd(self,df,cols=[]):
        if len(cols) == 0:
            cols = df.columns
        newdf = copy.deepcopy(df)
        self.class_index = {}
        for f in cols:
            uf = df[f].unique()
            self.class_index[f] = {}
            index = 1
            for item in uf:
                self.class_index[f][item] = index
                index += 1

            newdf[f] = df[f].apply(lambda d: self.update_label(f,d))
        return newdf

    def transform_pd(self,df,cols=[]):
        newdf = copy.deepcopy(df)
        if len(cols) == 0:
            cols = df.columns
        for f in cols:
            if f in self.class_index:
                newdf[f] = df[f].apply(lambda d: self.update_label(f,d))
        return newdf

    def update_label(self,f,x):
        try:
            return self.class_index[f][x]
        except:
            self.class_index[f][x] = max(self.class_index[f].values())+1
            return self.class_index[f][x]

In [ ]:
def min_max_scaler(df, numerical_columns, scaler_path):
    df = df.loc[:, numerical_columns].copy()
    scalers = {}
    mm_sclaers = []
    # Set Minmax scaler
    if not os.path.exists(scaler_path):
        print('Make a New Min Max Scaler')

        # Scale for each Numeric Columns
        for col in numerical_columns:
            mm_sclaer = MinMaxScaler()

            # Fit the scaler
            mm_sclaer.fit(df.loc[:, [col]])
            mm_sclaers.append(mm_sclaer)

            # Normalize each value between 0~1 using the scaler
            df.loc[:, col] = mm_sclaer.transform(df.loc[:, [col]])
            scalers[col] = mm_sclaer

        # Store the encoder
        f = open(scaler_path, 'wb')
        pickle.dump(scalers, f)

    else:
        print('Normalize with existing Min Max Scaler')
        # load scalers
        f = open(scaler_path, 'rb')
        scalers = pickle.load(f)

        # Normalize each value between 0~1 using loaded scaler
        for col in numerical_columns:
            mm_scaler = scalers[col]
            df.loc[:, col] = mm_scaler.transform(df.loc[:, [col]])

    return df, scalers

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!ls

drive  sample_data


In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!ls /content/drive/My\ Drive/

'0000 Pending case summery.xlsx'
 Books
 Certificates
 Classroom
'Colab Notebooks'
 Communication.xlsx
'Copy of Template Article Proceeding_4th WCO RTC Indonesia International Conference 2024.docx'
'Copy of Template Article Proceeding_4th WCO RTC Indonesia International Conference 2024.gdoc'
'Current Affairs'
'Format to Submit Independent Studies Titles.doc'
 IELTS
 IMG_20201221_160223.jpg
 IMG-20220318-WA0000.jpg
 KAndy.xlsx
 Magammana.xlsx
 MSC
'M.Sc. Data Science'
 Project.rar
'Recovery Details 2022 OIC.gsheet'
'RESEARCH M.M.V.N. Nandasena Final Report.pdf'
 RTF
'Vihanga Nikeshana'
 VISA
'VISA SG and MY'


In [ ]:
!ls /content/drive/My\ Drive/MSC/

Data_Mining  ML


In [ ]:
!ls /content/drive/My\ Drive/MSC/ML

df_syn_eng.csv


In [ ]:
file_path = '/content/drive/My Drive/MSC/ML/df_syn_eng.csv'
df = pd.read_csv(file_path)

In [ ]:
df.head(12)

Declaration ID        Date  Office ID Process Type  Import Type  \
0         97061800  2020-01-01         30            B           11   
1         85945189  2020-01-01         40            B           11   
2         77854751  2020-01-01         20            B           11   
3         46925060  2020-01-01         40            B           11   
4         34131149  2020-01-01         20            B           11   
5         38459034  2020-01-01         40            B           11   
6         69602841  2020-01-01         40            B           11   
7         94070719  2020-01-01         10            B           80   
8         28079013  2020-01-01         13            B           11   
9         66853714  2020-01-01         30            B           11   
10        19246049  2020-01-01         20            B           11   
11        14879976  2020-01-01         30            B           11   

    Import Use  Payment Type  Mode of Transport Declarant ID Importer ID  ...  \
0           21            11                 10      ZZR1LT6     QLRUBN9  ...   
1           21            11                 40      SWF9S4X     7JD1S2X  ...   
2           21            11                 40      X4XT6P8     WI9O3I5  ...   
3           21            43                 40      K7LCQTZ     6LI9721  ...   
4           21            11                 10      1HMVIVH     RZ871V1  ...   
5           21            11                 10      OZB7KED     2EIESGV  ...   
6           21            43                 40      3BTA0QN     US268D0  ...   
7           26            11                 40      YEVMXEV     EAPRJGG  ...   
8           21            11                 40      MP58TZN     OQGTRGJ  ...   
9           21            11                 40      P0PG8TK     FI22XB4  ...   
10          21            11                 10      TLY6XIA     NM5XD6W  ...   
11          21            11                 10      H3RWZG0     KRQQAFQ  ...   

   HS6 Code Country of Departure  Country of Origin Tax Rate Tax Type  \
0    440890                   BE                 BE      0.0     FEU1   
1    690722                   CN                 CN      8.0        A   
2    620822                   CN                 CN      5.2     FCN1   
3    940350                   VN                 VN      0.0        C   
4     71080                   VN                 VN     27.0        A   
5    851712                   SG                 CN      0.0        C   
6    391000                   AZ                 DE      6.5        C   
7    220870                   CN                 CN     12.0     FCN1   
8    853670                   US                 US      0.0     FUS1   
9    940180                   CN                 CN      0.0        C   
10   220429                   IT                 IT     15.0        F   
11   851762                   CN                 CN      0.0        C   

    Country of Origin Indicator Net Mass  Item Price  Fraud  Critical Fraud  
0                             G    108.0   372254.40      0               0  
1                             Y  11352.0   375751.20      0               0  
2                             B     20.7     5353.02      0               0  
3                             Y   9218.0  1477645.40      0               0  
4                             B  24000.0  6364800.00      0               0  
5                             S      2.0    29078.80      1               1  
6                             E     25.0    10765.00      0               0  
7                             Y    500.0    78900.00      0               0  
8                             B      0.2      241.58      0               0  
9                             G   2053.0     2463.60      0               0  
10                            E      7.7     3151.61      0               0  
11                            S   1057.2   153294.00      0               0  

[12 rows x 22 columns]

In [ ]:
df.shape

(54000, 22)

In [ ]:
df.isnull().sum()

Declaration ID                     0
Date                               0
Office ID                          0
Process Type                       0
Import Type                        0
Import Use                         0
Payment Type                       0
Mode of Transport                  0
Declarant ID                       0
Importer ID                        0
Seller ID                       5201
Courier ID                     44462
HS6 Code                           0
Country of Departure               0
Country of Origin                  0
Tax Rate                           0
Tax Type                           0
Country of Origin Indicator        0
Net Mass                           0
Item Price                         0
Fraud                              0
Critical Fraud                     0
dtype: int64

In this data set there are 54000 rows. But in the column ''Seller ID'' there are 5201 null values and ''courier id'' 44462 null values. Both of them are crusial details. since there are 54000 rows remove null value of the seller id rows and removed the entire ''courier id '' column.


In [ ]:
df = df.dropna(subset=['Seller ID'])

In [ ]:
df = df.drop(columns=['Courier ID'])

In [ ]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 48799 entries, 0 to 53999
Data columns (total 21 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Declaration ID               48799 non-null  int64  
 1   Date                         48799 non-null  object 
 2   Office ID                    48799 non-null  int64  
 3   Process Type                 48799 non-null  object 
 4   Import Type                  48799 non-null  int64  
 5   Import Use                   48799 non-null  int64  
 6   Payment Type                 48799 non-null  int64  
 7   Mode of Transport            48799 non-null  int64  
 8   Declarant ID                 48799 non-null  object 
 9   Importer ID                  48799 non-null  object 
 10  Seller ID                    48799 non-null  object 
 11  HS6 Code                     48799 non-null  int64  
 12  Country of Departure         48799 non-null  object 
 13  Country of Origin    

In [ ]:
df.isnull().sum()

Declaration ID                 0
Date                           0
Office ID                      0
Process Type                   0
Import Type                    0
Import Use                     0
Payment Type                   0
Mode of Transport              0
Declarant ID                   0
Importer ID                    0
Seller ID                      0
HS6 Code                       0
Country of Departure           0
Country of Origin              0
Tax Rate                       0
Tax Type                       0
Country of Origin Indicator    0
Net Mass                       0
Item Price                     0
Fraud                          0
Critical Fraud                 0
dtype: int64

In [ ]:
df.nunique()

Declaration ID                 48799
Date                             547
Office ID                         43
Process Type                       3
Import Type                       28
Import Use                        14
Payment Type                      10
Mode of Transport                  6
Declarant ID                    1238
Importer ID                    14297
Seller ID                      12201
HS6 Code                        2474
Country of Departure             102
Country of Origin                118
Tax Rate                         118
Tax Type                          48
Country of Origin Indicator        6
Net Mass                        5633
Item Price                     15617
Fraud                              2
Critical Fraud                     3
dtype: int64

In [ ]:

column_count = df.shape[1]
print(f"The DataFrame has {column_count} columns.")


The DataFrame has 21 columns.


In [ ]:
fraud_total_count = df['Declaration ID'].count()
fraud_value_counts = df['Declaration ID'].value_counts()
print(fraud_value_counts)

fraud_total_count = df['Fraud'].count()
fraud_value_counts = df['Fraud'].value_counts()
print(fraud_value_counts)

fraud_total_count = df['Critical Fraud'].count()
fraud_value_counts = df['Critical Fraud'].value_counts()
print(fraud_value_counts)

Declaration ID
97061800    1
49189825    1
42732342    1
70502727    1
87142567    1
           ..
28078133    1
11076640    1
33425392    1
34617551    1
24376062    1
Name: count, Length: 48799, dtype: int64
Fraud
0    38212
1    10587
Name: count, dtype: int64
Critical Fraud
0    38212
1    10097
2      490
Name: count, dtype: int64


In [ ]:
column_types = df.dtypes
print(column_types)

Declaration ID                   int64
Date                            object
Office ID                        int64
Process Type                    object
Import Type                      int64
Import Use                       int64
Payment Type                     int64
Mode of Transport                int64
Declarant ID                    object
Importer ID                     object
Seller ID                       object
HS6 Code                         int64
Country of Departure            object
Country of Origin               object
Tax Rate                       float64
Tax Type                        object
Country of Origin Indicator     object
Net Mass                       float64
Item Price                     float64
Fraud                            int64
Critical Fraud                   int64
dtype: object


In [ ]:
print(df.head())

   Declaration ID        Date  Office ID Process Type  Import Type  \
0        97061800  2020-01-01         30            B           11   
1        85945189  2020-01-01         40            B           11   
2        77854751  2020-01-01         20            B           11   
3        46925060  2020-01-01         40            B           11   
5        38459034  2020-01-01         40            B           11   

   Import Use  Payment Type  Mode of Transport Declarant ID Importer ID  ...  \
0          21            11                 10      ZZR1LT6     QLRUBN9  ...   
1          21            11                 40      SWF9S4X     7JD1S2X  ...   
2          21            11                 40      X4XT6P8     WI9O3I5  ...   
3          21            43                 40      K7LCQTZ     6LI9721  ...   
5          21            11                 10      OZB7KED     2EIESGV  ...   

  HS6 Code  Country of Departure Country of Origin Tax Rate  Tax Type  \
0   440890               

In [ ]:
df = df.drop(columns=['Date'])

In [ ]:
print(df.head())

   Declaration ID  Office ID Process Type  Import Type  Import Use  \
0        97061800         30            B           11          21   
1        85945189         40            B           11          21   
2        77854751         20            B           11          21   
3        46925060         40            B           11          21   
5        38459034         40            B           11          21   

   Payment Type  Mode of Transport Declarant ID Importer ID Seller ID  \
0            11                 10      ZZR1LT6     QLRUBN9   0VKY2BR   
1            11                 40      SWF9S4X     7JD1S2X   8WDKMC6   
2            11                 40      X4XT6P8     WI9O3I5   4DT3246   
3            43                 40      K7LCQTZ     6LI9721   PKUOG2P   
5            11                 10      OZB7KED     2EIESGV   WLTGD61   

   HS6 Code Country of Departure Country of Origin  Tax Rate Tax Type  \
0    440890                   BE                BE       0.0     FE

In [ ]:
import pandas as pd
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# Load your dataset (if not already loaded)
# df = pd.read_csv('/path/to/your/file.csv')

# Convert categorical columns to numeric using one-hot encoding
df = pd.get_dummies(df, columns=['Process Type', 'Declarant ID', 'Importer ID', 'Seller ID', 'Country of Departure', 'Country of Origin', 'Tax Type', 'Country of Origin Indicator'])

In [ ]:
# Split data into features and target variables
X = df.drop(columns=['Fraud', 'Critical Fraud'])
y_fraud = df['Fraud']
y_critical_fraud = df['Critical Fraud']

In [ ]:
# Train-test split
X_train, X_test, y_train_fraud, y_test_fraud = train_test_split(X, y_fraud, test_size=0.3, random_state=42)
X_train, X_test, y_train_critical, y_test_critical = train_test_split(X, y_critical_fraud, test_size=0.3, random_state=42)

In [ ]:
# Initialize and train XGBoost model for Fraud
model_fraud = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
model_fraud.fit(X_train, y_train_fraud)

In [ ]:
# Make predictions and evaluate
y_pred_fraud = model_fraud.predict(X_test)
print("Fraud Classification Report:\n", classification_report(y_test_fraud, y_pred_fraud))

# Initialize and train XGBoost model for Critical Fraud
model_critical = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
model_critical.fit(X_train, y_train_critical)

# Make predictions and evaluate
y_pred_critical = model_critical.predict(X_test)
print("Critical Fraud Classification Report:\n", classification_report(y_test_critical, y_pred_critical))

In [ ]:
import pandas as pd
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# Load your dataset (if not already loaded)
# df = pd.read_csv('/path/to/your/file.csv')

# Convert categorical columns to numeric using one-hot encoding
df = pd.get_dummies(df, columns=['Process Type', 'Declarant ID', 'Importer ID', 'Seller ID', 'Country of Departure', 'Country of Origin', 'Tax Type', 'Country of Origin Indicator'])

# Check the resulting DataFrame
# print(df.head())

# Split data into features and target variables
X = df.drop(columns=['Fraud', 'Critical Fraud'])
y_fraud = df['Fraud']
y_critical_fraud = df['Critical Fraud']

# Train-test split
X_train, X_test, y_train_fraud, y_test_fraud = train_test_split(X, y_fraud, test_size=0.3, random_state=42)
X_train, X_test, y_train_critical, y_test_critical = train_test_split(X, y_critical_fraud, test_size=0.3, random_state=42)

# Initialize and train XGBoost model for Fraud
model_fraud = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
model_fraud.fit(X_train, y_train_fraud)

# Make predictions and evaluate
y_pred_fraud = model_fraud.predict(X_test)
print("Fraud Classification Report:\n", classification_report(y_test_fraud, y_pred_fraud))

# Initialize and train XGBoost model for Critical Fraud
model_critical = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')
model_critical.fit(X_train, y_train_critical)

# Make predictions and evaluate
y_pred_critical = model_critical.predict(X_test)
print("Critical Fraud Classification Report:\n", classification_report(y_test_critical, y_pred_critical))


KeyError: "None of [Index(['Process Type', 'Declarant ID', 'Importer ID', 'Seller ID',\n       'Country of Departure', 'Country of Origin', 'Tax Type',\n       'Country of Origin Indicator'],\n      dtype='object')] are in the [columns]"

In [ ]:
import joblib

# Save the XGBoost model for Fraud
joblib.dump(model_fraud, 'xgboost_fraud_model.pkl')

# Save the XGBoost model for Critical Fraud
joblib.dump(model_critical, 'xgboost_critical_fraud_model.pkl')


In [ ]:
import shap
import xgboost as xgb
import pandas as pd

# Assuming you have trained your XGBoost model and have your data prepared
# model_fraud and X_train are previously defined

# Create a SHAP explainer object
explainer = shap.Explainer(model_fraud, X_train)

# Compute SHAP values for the test set
shap_values = explainer(X_test)

# Summary plot of SHAP values
shap.summary_plot(shap_values, X_test)

# Force plot for a single instance (optional, requires Jupyter notebook or similar environment)
# shap.force_plot(explainer.expected_value, shap_values[0], X_test.iloc[0])
